crates/doctor/src/manifest/version.rs

```diff
@@ -0,0 +1,122 @@
+use anyhow::{Context, Result};
+use async_trait::async_trait;
+use serde::Deserialize;
+use toml::Value;
+use toml_edit::{de::from_document, Document, Item};
+
+use crate::{Diagnose, Diagnosis, PatientApp, Treatment};
+
+use super::ManifestTreatment;
+
+const SPIN_MANIFEST_VERSION: &str = "spin_manifest_version";
+const SPIN_VERSION: &str = "spin_version";
+
+/// VersionDiagnosis detects problems with the app manifest version field.
+#[derive(Debug)]
+pub enum VersionDiagnosis {
+    /// Missing any known version key
+    MissingVersion,
+    /// Using old spin_version key
+    OldVersionKey,
+    /// Wrong version value
+    WrongValue(Value),
+}
+
+#[derive(Debug, Deserialize)]
+struct VersionProbe {
+    spin_manifest_version: Option<Value>,
+    spin_version: Option<Value>,
+}
+
+#[async_trait]
+impl Diagnose for VersionDiagnosis {
+    async fn diagnose(patient: &PatientApp) -> Result<Vec<Self>> {
+        let doc = &patient.manifest_doc;
+        let test: VersionProbe =
+            from_document(doc.clone()).context("failed to decode VersionTest")?;
+
+        let mut diags = vec![];
+        if test.spin_version.is_some() {
+            diags.push(Self::OldVersionKey);
+        }
+        if let Some(value) = test.spin_manifest_version.or(test.spin_version) {
+            if value.as_str() != Some("1") {
+                diags.push(Self::WrongValue(value))
+            }
+        } else {
+            diags.push(Self::MissingVersion);
+        }
+        Ok(diags)
+    }
+}
+
+impl Diagnosis for VersionDiagnosis {
+    fn description(&self) -> String {
+        match self {
+            Self::MissingVersion => "Manifest missing 'spin_manifest_version' key".into(),
+            Self::OldVersionKey => "Manifest using old 'spin_version' key".into(),
+            Self::WrongValue(val) => {
+                format!(r#"Manifest 'spin_manifest_version' must be "1", not {val}"#)
+            }
+        }
+    }
+
+    fn is_critical(&self) -> bool {
+        !matches!(self, Self::OldVersionKey)
+    }
+
+    fn treatment(&self) -> Option<&dyn Treatment> {
+        Some(self)
+    }
+}
+
+#[async_trait]
+impl ManifestTreatment for VersionDiagnosis {
+    async fn treat_manifest(&self, doc: &mut Document) -> anyhow::Result<()> {
```

Comment:
Hmm... I was thinking the issues should be separate so that a user can opt-in to the separate treatments, but now that I think of it the treatment fixes both problems unconditionally anyway... I'll change it one way or another.